### PR TITLE
[CodeComplete] Make SourceKit::CodeCompletion::Completion store a reference to the underlying swift result instead of extending that type

### PR DESCRIPTION
**Pull request by ahoppen**
Previously, when creating a `SourceKit::CodeCompletion::Completion`, we needed to copy all fields from the underlying `SwiftResult` (aka `swift::ide::CodeCompletionResult`). The arena in which the `SwiftResult` was allocated still needed to be kept alive for the references stored in the `SwiftResult`.

To avoid this unnecessary copy, make `SourceKit::CodeCompletion::Completion` store a reference to the underlying `SwiftResult`.

While making this change, I also noticed that the `innerSink` created by `transformAndForwardResults` did not adopt the sink of the outer results, but was still referencing e.g. strings stored in it. The first commit in this PR makes sure that we retain outer sink from the inner sink.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test Linux

File: lib/IDE/CompletionInstance.cpp [hunk 1]
```diff
@@ -709,9 +709,8 @@ void swift::ide::CompletionInstance::codeComplete(
           CancellationFlag->load(std::memory_order_relaxed)) {
         Callback(ResultType::cancelled());
       } else {
-        MutableArrayRef<CodeCompletionResult *> Results = context.takeResults();
         assert(SwiftContext.swiftASTContext);
-        Callback(ResultType::success({Results, SwiftContext}));
+        Callback(ResultType::success({context.getResultSink(), SwiftContext}));
```
Comment (rintaro) [inline lib/IDE/CompletionInstance.cpp:713]: This copies `CodeCompletionResultSink` object which has `std::vector<CodeCompletionResult *> Results` etc. Copying it sounds heavy. Do we really want that? If so could you explain why?

Comment (ahoppen) [reply lib/IDE/CompletionInstance.cpp:713]: Good catch `CodeCompleteResult` should store `CodeCompletionResultSink` by reference. Updated it.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test Linux

**Review by rintaro — APPROVED** [review]

**Pull request merged**